json.Compact has an incompatible behavior

I have encountered an incompatible behavior.
https://gist.github.com/orisano/8aa749cd50fece0a61694ee60b87a208

## Source
```go
package main

import (
	"bytes"
	"testing"
	
	stdjson "encoding/json"
	gojson "github.com/goccy/go-json"
)

func TestCompact(t *testing.T) {
	src := []byte(`"\`)
	if err := stdjson.Compact(bytes.NewBuffer(nil), src); err != nil {
		t.Error(err)
	}
	if err := gojson.Compact(bytes.NewBuffer(nil), src); err != nil {
		t.Error(err)
	}
}
```

## Result
```
--- FAIL: TestCompact (0.00s)
    compact_test.go:14: invalid character ' ' in string escape code
panic: runtime error: index out of range [2] with length 2 [recovered]
	panic: runtime error: index out of range [2] with length 2

goroutine 49 [running]:
testing.tRunner.func1.1(0x1144440, 0xc0006a8020)
	/Users/orisano/sdk/go1.15.6/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc000682480)
	/Users/orisano/sdk/go1.15.6/src/testing/testing.go:1075 +0x41a
panic(0x1144440, 0xc0006a8020)
	/Users/orisano/sdk/go1.15.6/src/runtime/panic.go:969 +0x1b9
github.com/goccy/go-json.compact(0xc000066f20, 0xc000066f0e, 0x2, 0x2, 0x106b300, 0x1197e80, 0xc00069c020)
	/Users/orisano/go/pkg/mod/github.com/goccy/go-json@v0.4.3/compact.go:34 +0x469
github.com/goccy/go-json.Compact(...)
	/Users/orisano/go/pkg/mod/github.com/goccy/go-json@v0.4.3/json.go:349
gojson.TestCompact(0xc000682480)
	/Users/orisano/works/gojson/compact_test.go:16 +0x125
testing.tRunner(0xc000682480, 0x11758b8)
	/Users/orisano/sdk/go1.15.6/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/Users/orisano/sdk/go1.15.6/src/testing/testing.go:1168 +0x2b3
exit status 2
FAIL	gojson	0.290s
```

go-json panicked.
Maybe cause is here.
https://github.com/goccy/go-json/blob/7ffe1ddb3521e9abb485d5c47cf265fb74bea347/compact.go#L32-L36

What is the correct behavior?

A similar bug in the same function can be triggered by marshaling a raw.Message:

```
package main

import (
        "fmt"

        json "github.com/goccy/go-json"
//      "encoding/json"
)

type Foo struct {
        A json.RawMessage
}

func main() {
        input := []byte(`{"a":"\\""}`)
        foo := &Foo{}
        err := json.Unmarshal(input, foo)
        fmt.Println(foo, err)

        output, err := json.Marshal(foo)
        fmt.Println(output, err)
}
```

Which produces:
```
$ go run -trimpath main.go
&{A:[34 92 92 34 34]} <nil>
panic: runtime error: index out of range [5] with length 5

goroutine 1 [running]:
github.com/goccy/go-json/internal/encoder.Compact(0xc00007ce78, 0xc0000160b0, 0x5, 0x8, 0x1, 0x0, 0x0)
	github.com/goccy/go-json@v0.4.11/internal/encoder/compact.go:25 +0x55f
github.com/goccy/go-json/internal/encoder.AppendMarshalJSON(0xc000496500, 0xc000492000, 0x5, 0x400, 0x11a67a0, 0xc00000c060, 0x1, 0x0, 0x0, 0x0, ...)
	github.com/goccy/go-json@v0.4.11/internal/encoder/encoder.go:438 +0x3ca
github.com/goccy/go-json/internal/encoder/vm_escaped.Run(0xc000494000, 0xc000492000, 0x1, 0x400, 0xc00000c0d8, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/goccy/go-json@v0.4.11/internal/encoder/vm_escaped/vm.go:2721 +0x11145
github.com/goccy/go-json.encodeRunCode(0xc000494000, 0xc000492000, 0x0, 0x400, 0xc00000c0d8, 0x1, 0xc000492400, 0xc000492000, 0xc00007ddd0, 0x107852a, ...)
	github.com/goccy/go-json@v0.4.11/encode.go:281 +0x7e
github.com/goccy/go-json.encode(0xc000494000, 0x119da00, 0xc00000c060, 0x1, 0x10783fc, 0x12be100, 0xc000490000, 0x0, 0xc00007de48)
	github.com/goccy/go-json@v0.4.11/encode.go:211 +0x215
github.com/goccy/go-json.marshal(0x119da00, 0xc00000c060, 0x1, 0x0, 0x0, 0x1c, 0xc000060dd0, 0x0)
	github.com/goccy/go-json@v0.4.11/encode.go:135 +0x85
github.com/goccy/go-json.MarshalWithOption(0x119da00, 0xc00000c060, 0x0, 0x0, 0x0, 0x2, 0x2, 0x1c, 0x0, 0x0)
	github.com/goccy/go-json@v0.4.11/json.go:172 +0x76
github.com/goccy/go-json.Marshal(...)
	github.com/goccy/go-json@v0.4.11/json.go:158
main.main()
	command-line-arguments/main.go:20 +0x167
exit status 2
```

--- a/benchmarks/bench_test.go
+++ b/benchmarks/bench_test.go
@@ -581,3 +581,63 @@ func BenchmarkUnmapped(b *testing.B) {
 		}
 	})
 }
+
+func Benchmark_Compact_EncodingJson(b *testing.B) {
+	b.ReportAllocs()
+	if codeJSON == nil {
+		b.StopTimer()
+		codeInit()
+		b.StartTimer()
+	}
+	for i := 0; i < b.N; i++ {
+		var buf bytes.Buffer
+		if err := stdjson.Compact(&buf, codeJSON); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func Benchmark_Compact_GoJson(b *testing.B) {
+	b.ReportAllocs()
+	if codeJSON == nil {
+		b.StopTimer()
+		codeInit()
+		b.StartTimer()
+	}
+	for i := 0; i < b.N; i++ {
+		var buf bytes.Buffer
+		if err := json.Compact(&buf, codeJSON); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func Benchmark_Indent_EncodingJson(b *testing.B) {
+	b.ReportAllocs()
+	if codeJSON == nil {
+		b.StopTimer()
+		codeInit()
+		b.StartTimer()
+	}
+	for i := 0; i < b.N; i++ {
+		var buf bytes.Buffer
+		if err := stdjson.Indent(&buf, codeJSON, "-", " "); err != nil {
+			b.Fatal(err)
+		}
+	}
+}
+
+func Benchmark_Indent_GoJson(b *testing.B) {
+	b.ReportAllocs()
+	if codeJSON == nil {
+		b.StopTimer()
+		codeInit()
+		b.StartTimer()
+	}
+	for i := 0; i < b.N; i++ {
+		var buf bytes.Buffer
+		if err := json.Indent(&buf, codeJSON, "-", " "); err != nil {
+			b.Fatal(err)
+		}
+	}
+}

--- a/internal/encoder/compact.go
+++ b/internal/encoder/compact.go
@@ -7,7 +7,6 @@ import (
 	"unsafe"
 
 	"github.com/goccy/go-json/internal/errors"
-	"github.com/goccy/go-json/internal/runtime"
 )
 
 var (
@@ -29,8 +28,11 @@ func Compact(buf *bytes.Buffer, src []byte, escape bool) error {
 	if len(src) == 0 {
 		return errors.ErrUnexpectedEndOfJSON("", 0)
 	}
-	dst := make([]byte, 0, len(src))
-	dst, err := compact(dst, src, escape)
+	buf.Grow(len(src))
+	dst := buf.Bytes()
+	newSrc := make([]byte, len(src)+1) // append nul byte to the end
+	copy(newSrc, src)
+	dst, err := compact(dst, newSrc, escape)
 	if err != nil {
 		return err
 	}
@@ -41,12 +43,10 @@ func Compact(buf *bytes.Buffer, src []byte, escape bool) error {
 }
 
 func compact(dst, src []byte, escape bool) ([]byte, error) {
-	src = append(src, nul)
 	buf, _, err := compactValue(dst, src, 0, escape)
 	if err != nil {
 		return nil, err
 	}
-	(*runtime.SliceHeader)(unsafe.Pointer(&src)).Len--
 	return buf, nil
 }
 

--- a/internal/encoder/compact.go
+++ b/internal/encoder/compact.go
@@ -2,55 +2,259 @@ package encoder
 
 import (
 	"bytes"
+	"fmt"
+	"strconv"
+	"unsafe"
 
 	"github.com/goccy/go-json/internal/errors"
+	"github.com/goccy/go-json/internal/runtime"
 )
 
-func Compact(dst *bytes.Buffer, src []byte, escape bool) error {
+var (
+	isWhiteSpace = [256]bool{
+		' ':  true,
+		'\n': true,
+		'\t': true,
+		'\r': true,
+	}
+	isHTMLEscapeChar = [256]bool{
+		'<': true,
+		'>': true,
+		'&': true,
+	}
+	nul = byte('\000')
+)
+
+func Compact(buf *bytes.Buffer, src []byte, escape bool) error {
 	if len(src) == 0 {
 		return errors.ErrUnexpectedEndOfJSON("", 0)
 	}
-	length := len(src)
-	for cursor := 0; cursor < length; cursor++ {
-		c := src[cursor]
-		switch c {
-		case ' ', '\t', '\n', '\r':
-			continue
-		case '"':
-			if err := dst.WriteByte(c); err != nil {
-				return err
-			}
-			for {
-				cursor++
-				c := src[cursor]
-				if escape && (c == '<' || c == '>' || c == '&') {
-					if _, err := dst.WriteString(`\u00`); err != nil {
-						return err
-					}
-					if _, err := dst.Write([]byte{hex[c>>4], hex[c&0xF]}); err != nil {
-						return err
-					}
-				} else if err := dst.WriteByte(c); err != nil {
-					return err
-				}
-				switch c {
-				case '\\':
-					cursor++
-					if err := dst.WriteByte(src[cursor]); err != nil {
-						return err
-					}
-				case '"':
-					goto LOOP_END
-				case '\000':
-					return errors.ErrUnexpectedEndOfJSON("string", int64(length))
-				}
-			}
-		default:
-			if err := dst.WriteByte(c); err != nil {
-				return err
-			}
-		}
-	LOOP_END:
+	dst := make([]byte, 0, len(src))
+	dst, err := compact(dst, src, escape)
+	if err != nil {
+		return err
+	}
+	if _, err := buf.Write(dst); err != nil {
+		return err
 	}
 	return nil
+}
+
+func compact(dst, src []byte, escape bool) ([]byte, error) {
+	src = append(src, nul)
+	buf, _, err := compactValue(dst, src, 0, escape)
+	if err != nil {
+		return nil, err
+	}
+	(*runtime.SliceHeader)(unsafe.Pointer(&src)).Len--
+	return buf, nil
+}
+
+func skipWhiteSpace(buf []byte, cursor int64) int64 {
+LOOP:
+	if isWhiteSpace[buf[cursor]] {
+		cursor++
+		goto LOOP
+	}
+	return cursor
+}
+
+func compactValue(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
+	for {
+		switch src[cursor] {
+		case ' ', '\t', '\n', '\r':
+			cursor++
+			continue
+		case '{':
+			return compactObject(dst, src, cursor, escape)
+		case '}':
+			return nil, 0, errors.ErrSyntax("unexpected character '}'", cursor)
+		case '[':
+			return compactArray(dst, src, cursor, escape)
+		case ']':
+			return nil, 0, errors.ErrSyntax("unexpected character ']'", cursor)
+		case '"':
+			return compactString(dst, src, cursor, escape)
+		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+			return compactNumber(dst, src, cursor)
+		case 't':
+			return compactTrue(dst, src, cursor)
+		case 'f':
+			return compactFalse(dst, src, cursor)
+		case 'n':
+			return compactNull(dst, src, cursor)
+		default:
+			return nil, 0, errors.ErrSyntax(fmt.Sprintf("unexpected character '%s'", src[cursor]), cursor)
+		}
+	}
+}
+
+func compactObject(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
+	switch src[cursor] {
+	case 'n':
+		return compactNull(dst, src, cursor)
+	case '{':
+		dst = append(dst, '{')
+	default:
+		return nil, 0, errors.ErrExpected("expected { character for object value", cursor)
+	}
+	cursor = skipWhiteSpace(src, cursor+1)
+	if src[cursor] == '}' {
+		dst = append(dst, '}')
+		return dst, cursor + 1, nil
+	}
+	var err error
+	for {
+		cursor = skipWhiteSpace(src, cursor)
+		dst, cursor, err = compactString(dst, src, cursor, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		if src[cursor] != ':' {
+			return nil, 0, errors.ErrExpected("colon after object key", cursor)
+		}
+		dst = append(dst, ':')
+		dst, cursor, err = compactValue(dst, src, cursor+1, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		switch src[cursor] {
+		case '}':
+			dst = append(dst, '}')
+			cursor++
+			return dst, cursor, nil
+		case ',':
+			dst = append(dst, ',')
+		default:
+			return nil, 0, errors.ErrExpected("comma after object value", cursor)
+		}
+		cursor++
+	}
+}
+
+func compactArray(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
+	switch src[cursor] {
+	case 'n':
+		return compactNull(dst, src, cursor)
+	case '[':
+		dst = append(dst, '[')
+	default:
+		return nil, 0, errors.ErrExpected("expected [ character for array value", cursor)
+	}
+	cursor = skipWhiteSpace(src, cursor+1)
+	if src[cursor] == ']' {
+		dst = append(dst, ']')
+		return dst, cursor + 1, nil
+	}
+	var err error
+	for {
+		dst, cursor, err = compactValue(dst, src, cursor, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		switch src[cursor] {
+		case ']':
+			dst = append(dst, ']')
+			cursor++
+			return dst, cursor, nil
+		case ',':
+			dst = append(dst, ',')
+		default:
+			return nil, 0, errors.ErrExpected("comma after array value", cursor)
+		}
+		cursor++
+	}
+}
+
+func compactString(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
+	if src[cursor] != '"' {
+		return nil, 0, errors.ErrInvalidCharacter(src[cursor], "string", cursor)
+	}
+	start := cursor
+	for {
+		cursor++
+		c := src[cursor]
+		if escape {
+			if isHTMLEscapeChar[c] {
+				dst = append(dst, src[start:cursor]...)
+				dst = append(dst, `\u00`...)
+				dst = append(dst, hex[c>>4], hex[c&0xF])
+				start = cursor + 1
+			} else if c == 0xE2 && cursor+2 < int64(len(src)) && src[cursor+1] == 0x80 && src[cursor+2]&^1 == 0xA8 {
+				dst = append(dst, src[start:cursor]...)
+				dst = append(dst, `\u202`...)
+				dst = append(dst, hex[src[cursor+2]&0xF])
+				cursor += 2
+				start = cursor + 3
+			}
+		}
+		switch c {
+		case '\\':
+			cursor++
+			if src[cursor] == nul {
+				return nil, 0, errors.ErrUnexpectedEndOfJSON("string", int64(len(src)))
+			}
+		case '"':
+			cursor++
+			return append(dst, src[start:cursor]...), cursor, nil
+		case nul:
+			return nil, 0, errors.ErrUnexpectedEndOfJSON("string", int64(len(src)))
+		}
+	}
+}
+
+func compactNumber(dst, src []byte, cursor int64) ([]byte, int64, error) {
+	start := cursor
+	for {
+		cursor++
+		if floatTable[src[cursor]] {
+			continue
+		}
+		break
+	}
+	num := src[start:cursor]
+	if _, err := strconv.ParseFloat(*(*string)(unsafe.Pointer(&num)), 64); err != nil {
+		return nil, 0, err
+	}
+	dst = append(dst, num...)
+	return dst, cursor, nil
+}
+
+func compactTrue(dst, src []byte, cursor int64) ([]byte, int64, error) {
+	if cursor+3 >= int64(len(src)) {
+		return nil, 0, errors.ErrUnexpectedEndOfJSON("true", cursor)
+	}
+	if !bytes.Equal(src[cursor:cursor+4], []byte(`true`)) {
+		return nil, 0, errors.ErrInvalidCharacter(src[cursor], "true", cursor)
+	}
+	dst = append(dst, "true"...)
+	cursor += 4
+	return dst, cursor, nil
+}
+
+func compactFalse(dst, src []byte, cursor int64) ([]byte, int64, error) {
+	if cursor+4 >= int64(len(src)) {
+		return nil, 0, errors.ErrUnexpectedEndOfJSON("false", cursor)
+	}
+	if !bytes.Equal(src[cursor:cursor+5], []byte(`false`)) {
+		return nil, 0, errors.ErrInvalidCharacter(src[cursor], "false", cursor)
+	}
+	dst = append(dst, "false"...)
+	cursor += 5
+	return dst, cursor, nil
+}
+
+func compactNull(dst, src []byte, cursor int64) ([]byte, int64, error) {
+	if cursor+3 >= int64(len(src)) {
+		return nil, 0, errors.ErrUnexpectedEndOfJSON("null", cursor)
+	}
+	if !bytes.Equal(src[cursor:cursor+4], []byte(`null`)) {
+		return nil, 0, errors.ErrInvalidCharacter(src[cursor], "null", cursor)
+	}
+	dst = append(dst, "null"...)
+	cursor += 4
+	return dst, cursor, nil
 }

--- a/internal/encoder/compact.go
+++ b/internal/encoder/compact.go
@@ -43,11 +43,30 @@ func Compact(buf *bytes.Buffer, src []byte, escape bool) error {
 }
 
 func compact(dst, src []byte, escape bool) ([]byte, error) {
-	buf, _, err := compactValue(dst, src, 0, escape)
+	buf, cursor, err := compactValue(dst, src, 0, escape)
 	if err != nil {
 		return nil, err
 	}
+	if err := validateEndBuf(src, cursor); err != nil {
+		return nil, err
+	}
 	return buf, nil
+}
+
+func validateEndBuf(src []byte, cursor int64) error {
+	for {
+		switch src[cursor] {
+		case ' ', '\t', '\n', '\r':
+			cursor++
+			continue
+		case nul:
+			return nil
+		}
+		return errors.ErrSyntax(
+			fmt.Sprintf("invalid character '%c' after top-level value", src[cursor]),
+			cursor+1,
+		)
+	}
 }
 
 func skipWhiteSpace(buf []byte, cursor int64) int64 {
@@ -90,12 +109,9 @@ func compactValue(dst, src []byte, cursor int64, escape bool) ([]byte, int64, er
 }
 
 func compactObject(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
-	switch src[cursor] {
-	case 'n':
-		return compactNull(dst, src, cursor)
-	case '{':
+	if src[cursor] == '{' {
 		dst = append(dst, '{')
-	default:
+	} else {
 		return nil, 0, errors.ErrExpected("expected { character for object value", cursor)
 	}
 	cursor = skipWhiteSpace(src, cursor+1)
@@ -135,12 +151,9 @@ func compactObject(dst, src []byte, cursor int64, escape bool) ([]byte, int64, e
 }
 
 func compactArray(dst, src []byte, cursor int64, escape bool) ([]byte, int64, error) {
-	switch src[cursor] {
-	case 'n':
-		return compactNull(dst, src, cursor)
-	case '[':
+	if src[cursor] == '[' {
 		dst = append(dst, '[')
-	default:
+	} else {
 		return nil, 0, errors.ErrExpected("expected [ character for array value", cursor)
 	}
 	cursor = skipWhiteSpace(src, cursor+1)

--- a/internal/encoder/compact.go
+++ b/internal/encoder/compact.go
@@ -84,7 +84,7 @@ func compactValue(dst, src []byte, cursor int64, escape bool) ([]byte, int64, er
 		case 'n':
 			return compactNull(dst, src, cursor)
 		default:
-			return nil, 0, errors.ErrSyntax(fmt.Sprintf("unexpected character '%s'", src[cursor]), cursor)
+			return nil, 0, errors.ErrSyntax(fmt.Sprintf("unexpected character '%c'", src[cursor]), cursor)
 		}
 	}
 }

--- a/internal/encoder/encoder.go
+++ b/internal/encoder/encoder.go
@@ -306,6 +306,7 @@ func MapLen(m unsafe.Pointer) int
 
 type RuntimeContext struct {
 	Buf        []byte
+	MarshalBuf []byte
 	Ptrs       []uintptr
 	KeepRefs   []unsafe.Pointer
 	SeenPtr    []uintptr
@@ -413,7 +414,7 @@ func AppendNumber(b []byte, n json.Number) ([]byte, error) {
 	return b, nil
 }
 
-func AppendMarshalJSON(code *Opcode, b []byte, v interface{}, escape bool) ([]byte, error) {
+func AppendMarshalJSON(ctx *RuntimeContext, code *Opcode, b []byte, v interface{}, escape bool) ([]byte, error) {
 	rv := reflect.ValueOf(v) // convert by dynamic interface type
 	if code.AddrForMarshaler {
 		if rv.CanAddr() {
@@ -433,10 +434,13 @@ func AppendMarshalJSON(code *Opcode, b []byte, v interface{}, escape bool) ([]by
 	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
-	compactedBuf, err := compact(b, bb, escape)
+	marshalBuf := ctx.MarshalBuf[:0]
+	marshalBuf = append(append(marshalBuf, bb...), nul)
+	compactedBuf, err := compact(b, marshalBuf, escape)
 	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
+	ctx.MarshalBuf = marshalBuf
 	return compactedBuf, nil
 }
 

--- a/internal/encoder/encoder.go
+++ b/internal/encoder/encoder.go
@@ -433,12 +433,11 @@ func AppendMarshalJSON(code *Opcode, b []byte, v interface{}, escape bool) ([]by
 	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
-	buf := bytes.NewBuffer(b)
-	// TODO: we should validate buffer with `compact`
-	if err := Compact(buf, bb, escape); err != nil {
+	compactedBuf, err := compact(b, bb, escape)
+	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
-	return buf.Bytes(), nil
+	return compactedBuf, nil
 }
 
 func AppendMarshalJSONIndent(ctx *RuntimeContext, code *Opcode, b []byte, v interface{}, indent int, escape bool) ([]byte, error) {

--- a/internal/encoder/encoder.go
+++ b/internal/encoder/encoder.go
@@ -464,20 +464,20 @@ func AppendMarshalJSONIndent(ctx *RuntimeContext, code *Opcode, b []byte, v inte
 	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
-	var compactBuf bytes.Buffer
-	if err := Compact(&compactBuf, bb, escape); err != nil {
-		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
-	}
-	var indentBuf bytes.Buffer
-	if err := Indent(
-		&indentBuf,
-		compactBuf.Bytes(),
+	marshalBuf := ctx.MarshalBuf[:0]
+	marshalBuf = append(append(marshalBuf, bb...), nul)
+	indentedBuf, err := doIndent(
+		b,
+		marshalBuf,
 		string(ctx.Prefix)+strings.Repeat(string(ctx.IndentStr), ctx.BaseIndent+indent),
 		string(ctx.IndentStr),
-	); err != nil {
+		escape,
+	)
+	if err != nil {
 		return nil, &errors.MarshalerError{Type: reflect.TypeOf(v), Err: err}
 	}
-	return append(b, indentBuf.Bytes()...), nil
+	ctx.MarshalBuf = marshalBuf
+	return indentedBuf, nil
 }
 
 func AppendMarshalText(code *Opcode, b []byte, v interface{}, escape bool) ([]byte, error) {

--- a/internal/encoder/indent.go
+++ b/internal/encoder/indent.go
@@ -2,111 +2,179 @@ package encoder
 
 import (
 	"bytes"
+	"fmt"
 
 	"github.com/goccy/go-json/internal/errors"
 )
 
-func Indent(dst *bytes.Buffer, src []byte, prefix, indentStr string) error {
-	length := int64(len(src))
-	indentNum := 0
-	indentBytes := []byte(indentStr)
-	for cursor := int64(0); cursor < length; cursor++ {
-		c := src[cursor]
-		switch c {
-		case ' ', '\t', '\n', '\r':
-			continue
-		case '"':
-			if err := dst.WriteByte(c); err != nil {
-				return err
-			}
-			for {
-				cursor++
-				if err := dst.WriteByte(src[cursor]); err != nil {
-					return err
-				}
-				switch src[cursor] {
-				case '\\':
-					cursor++
-					if err := dst.WriteByte(src[cursor]); err != nil {
-						return err
-					}
-				case '"':
-					goto LOOP_END
-				case '\000':
-					return errors.ErrUnexpectedEndOfJSON("string", length)
-				}
-			}
-		case '{':
-			if cursor+1 < length && src[cursor+1] == '}' {
-				if _, err := dst.Write([]byte{'{', '}'}); err != nil {
-					return err
-				}
-				cursor++
-			} else {
-				indentNum++
-				b := []byte{c, '\n'}
-				b = append(b, prefix...)
-				b = append(b, bytes.Repeat(indentBytes, indentNum)...)
-				if _, err := dst.Write(b); err != nil {
-					return err
-				}
-			}
-		case '}':
-			indentNum--
-			if indentNum < 0 {
-				return errors.ErrInvalidCharacter('}', "}", cursor)
-			}
-			b := []byte{'\n'}
-			b = append(b, prefix...)
-			b = append(b, bytes.Repeat(indentBytes, indentNum)...)
-			b = append(b, c)
-			if _, err := dst.Write(b); err != nil {
-				return err
-			}
-		case '[':
-			if cursor+1 < length && src[cursor+1] == ']' {
-				if _, err := dst.Write([]byte{'[', ']'}); err != nil {
-					return err
-				}
-				cursor++
-			} else {
-				indentNum++
-				b := []byte{c, '\n'}
-				b = append(b, prefix...)
-				b = append(b, bytes.Repeat(indentBytes, indentNum)...)
-				if _, err := dst.Write(b); err != nil {
-					return err
-				}
-			}
-		case ']':
-			indentNum--
-			if indentNum < 0 {
-				return errors.ErrInvalidCharacter(']', "]", cursor)
-			}
-			b := []byte{'\n'}
-			b = append(b, prefix...)
-			b = append(b, bytes.Repeat(indentBytes, indentNum)...)
-			b = append(b, c)
-			if _, err := dst.Write(b); err != nil {
-				return err
-			}
-		case ':':
-			if _, err := dst.Write([]byte{':', ' '}); err != nil {
-				return err
-			}
-		case ',':
-			b := []byte{',', '\n'}
-			b = append(b, prefix...)
-			b = append(b, bytes.Repeat(indentBytes, indentNum)...)
-			if _, err := dst.Write(b); err != nil {
-				return err
-			}
-		default:
-			if err := dst.WriteByte(c); err != nil {
-				return err
-			}
-		}
-	LOOP_END:
+func Indent(buf *bytes.Buffer, src []byte, prefix, indentStr string) error {
+	if len(src) == 0 {
+		return errors.ErrUnexpectedEndOfJSON("", 0)
+	}
+	buf.Grow(len(src))
+	dst := buf.Bytes()
+	newSrc := make([]byte, len(src)+1) // append nul byte to the end
+	copy(newSrc, src)
+	dst, err := doIndent(dst, newSrc, prefix, indentStr, false)
+	if err != nil {
+		return err
+	}
+	if _, err := buf.Write(dst); err != nil {
+		return err
 	}
 	return nil
+}
+
+func doIndent(dst, src []byte, prefix, indentStr string, escape bool) ([]byte, error) {
+	buf, _, err := indentValue(dst, src, 0, 0, []byte(prefix), []byte(indentStr), escape)
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+func indentValue(
+	dst []byte,
+	src []byte,
+	indentNum int,
+	cursor int64,
+	prefix []byte,
+	indentBytes []byte,
+	escape bool) ([]byte, int64, error) {
+	for {
+		switch src[cursor] {
+		case ' ', '\t', '\n', '\r':
+			cursor++
+			continue
+		case '{':
+			return indentObject(dst, src, indentNum, cursor, prefix, indentBytes, escape)
+		case '}':
+			return nil, 0, errors.ErrSyntax("unexpected character '}'", cursor)
+		case '[':
+			return indentArray(dst, src, indentNum, cursor, prefix, indentBytes, escape)
+		case ']':
+			return nil, 0, errors.ErrSyntax("unexpected character ']'", cursor)
+		case '"':
+			return compactString(dst, src, cursor, escape)
+		case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+			return compactNumber(dst, src, cursor)
+		case 't':
+			return compactTrue(dst, src, cursor)
+		case 'f':
+			return compactFalse(dst, src, cursor)
+		case 'n':
+			return compactNull(dst, src, cursor)
+		default:
+			return nil, 0, errors.ErrSyntax(fmt.Sprintf("unexpected character '%c'", src[cursor]), cursor)
+		}
+	}
+}
+
+func indentObject(
+	dst []byte,
+	src []byte,
+	indentNum int,
+	cursor int64,
+	prefix []byte,
+	indentBytes []byte,
+	escape bool) ([]byte, int64, error) {
+	switch src[cursor] {
+	case 'n':
+		return compactNull(dst, src, cursor)
+	case '{':
+		dst = append(dst, '{')
+	default:
+		return nil, 0, errors.ErrExpected("expected { character for object value", cursor)
+	}
+	cursor = skipWhiteSpace(src, cursor+1)
+	if src[cursor] == '}' {
+		dst = append(dst, '}')
+		return dst, cursor + 1, nil
+	}
+	indentNum++
+	var err error
+	for {
+		dst = append(append(append(dst, '\n'), prefix...), bytes.Repeat(indentBytes, indentNum)...)
+		cursor = skipWhiteSpace(src, cursor)
+		dst, cursor, err = compactString(dst, src, cursor, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		if src[cursor] != ':' {
+			return nil, 0, errors.ErrSyntax(
+				fmt.Sprintf("invalid character '%c' after object key", src[cursor]),
+				cursor+1,
+			)
+		}
+		dst = append(dst, ':', ' ')
+		dst, cursor, err = indentValue(dst, src, indentNum, cursor+1, prefix, indentBytes, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		switch src[cursor] {
+		case '}':
+			dst = append(append(append(dst, '\n'), prefix...), bytes.Repeat(indentBytes, indentNum-1)...)
+			dst = append(dst, '}')
+			cursor++
+			return dst, cursor, nil
+		case ',':
+			dst = append(dst, ',')
+		default:
+			return nil, 0, errors.ErrSyntax(
+				fmt.Sprintf("invalid character '%c' after object key:value pair", src[cursor]),
+				cursor+1,
+			)
+		}
+		cursor++
+	}
+}
+
+func indentArray(
+	dst []byte,
+	src []byte,
+	indentNum int,
+	cursor int64,
+	prefix []byte,
+	indentBytes []byte,
+	escape bool) ([]byte, int64, error) {
+	switch src[cursor] {
+	case 'n':
+		return compactNull(dst, src, cursor)
+	case '[':
+		dst = append(dst, '[')
+	default:
+		return nil, 0, errors.ErrExpected("expected [ character for array value", cursor)
+	}
+	cursor = skipWhiteSpace(src, cursor+1)
+	if src[cursor] == ']' {
+		dst = append(dst, ']')
+		return dst, cursor + 1, nil
+	}
+	indentNum++
+	var err error
+	for {
+		dst = append(append(append(dst, '\n'), prefix...), bytes.Repeat(indentBytes, indentNum)...)
+		dst, cursor, err = indentValue(dst, src, indentNum, cursor, prefix, indentBytes, escape)
+		if err != nil {
+			return nil, 0, err
+		}
+		cursor = skipWhiteSpace(src, cursor)
+		switch src[cursor] {
+		case ']':
+			dst = append(append(append(dst, '\n'), prefix...), bytes.Repeat(indentBytes, indentNum-1)...)
+			dst = append(dst, ']')
+			cursor++
+			return dst, cursor, nil
+		case ',':
+			dst = append(dst, ',')
+		default:
+			return nil, 0, errors.ErrSyntax(
+				fmt.Sprintf("invalid character '%c' after array value", src[cursor]),
+				cursor+1,
+			)
+		}
+		cursor++
+	}
 }

--- a/internal/encoder/indent.go
+++ b/internal/encoder/indent.go
@@ -26,8 +26,11 @@ func Indent(buf *bytes.Buffer, src []byte, prefix, indentStr string) error {
 }
 
 func doIndent(dst, src []byte, prefix, indentStr string, escape bool) ([]byte, error) {
-	buf, _, err := indentValue(dst, src, 0, 0, []byte(prefix), []byte(indentStr), escape)
+	buf, cursor, err := indentValue(dst, src, 0, 0, []byte(prefix), []byte(indentStr), escape)
 	if err != nil {
+		return nil, err
+	}
+	if err := validateEndBuf(src, cursor); err != nil {
 		return nil, err
 	}
 	return buf, nil
@@ -78,12 +81,9 @@ func indentObject(
 	prefix []byte,
 	indentBytes []byte,
 	escape bool) ([]byte, int64, error) {
-	switch src[cursor] {
-	case 'n':
-		return compactNull(dst, src, cursor)
-	case '{':
+	if src[cursor] == '{' {
 		dst = append(dst, '{')
-	default:
+	} else {
 		return nil, 0, errors.ErrExpected("expected { character for object value", cursor)
 	}
 	cursor = skipWhiteSpace(src, cursor+1)
@@ -139,12 +139,9 @@ func indentArray(
 	prefix []byte,
 	indentBytes []byte,
 	escape bool) ([]byte, int64, error) {
-	switch src[cursor] {
-	case 'n':
-		return compactNull(dst, src, cursor)
-	case '[':
+	if src[cursor] == '[' {
 		dst = append(dst, '[')
-	default:
+	} else {
 		return nil, 0, errors.ErrExpected("expected [ character for array value", cursor)
 	}
 	cursor = skipWhiteSpace(src, cursor+1)

--- a/internal/encoder/vm/vm.go
+++ b/internal/encoder/vm/vm.go
@@ -273,7 +273,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if code.IsNilableType && code.Indirect {
 				p = ptrToPtr(p)
 			}
-			bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+			bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 			if err != nil {
 				return nil, err
 			}
@@ -2718,7 +2718,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2762,7 +2762,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2807,7 +2807,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, iface, false)
+				bb, err := appendMarshalJSON(ctx, code, b, iface, false)
 				if err != nil {
 					return nil, err
 				}
@@ -2847,7 +2847,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2887,7 +2887,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3628,7 +3628,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3652,7 +3652,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				break
 			}
 			b = append(b, code.Key...)
-			bb, err := appendMarshalJSON(code, b, iface, false)
+			bb, err := appendMarshalJSON(ctx, code, b, iface, false)
 			if err != nil {
 				return nil, err
 			}
@@ -3665,7 +3665,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3678,7 +3678,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			p = ptrToNPtr(p+code.Offset, code.PtrNum)
 			if p != 0 {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}

--- a/internal/encoder/vm_debug/vm.go
+++ b/internal/encoder/vm_debug/vm.go
@@ -286,7 +286,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if code.IsNilableType && code.Indirect {
 				p = ptrToPtr(p)
 			}
-			bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+			bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 			if err != nil {
 				return nil, err
 			}
@@ -2731,7 +2731,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2775,7 +2775,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2820,7 +2820,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, iface, false)
+				bb, err := appendMarshalJSON(ctx, code, b, iface, false)
 				if err != nil {
 					return nil, err
 				}
@@ -2860,7 +2860,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -2900,7 +2900,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3641,7 +3641,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3665,7 +3665,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				break
 			}
 			b = append(b, code.Key...)
-			bb, err := appendMarshalJSON(code, b, iface, false)
+			bb, err := appendMarshalJSON(ctx, code, b, iface, false)
 			if err != nil {
 				return nil, err
 			}
@@ -3678,7 +3678,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}
@@ -3691,7 +3691,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			p = ptrToNPtr(p+code.Offset, code.PtrNum)
 			if p != 0 {
 				b = append(b, code.Key...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), false)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), false)
 				if err != nil {
 					return nil, err
 				}

--- a/internal/encoder/vm_escaped/vm.go
+++ b/internal/encoder/vm_escaped/vm.go
@@ -273,7 +273,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if code.IsNilableType && code.Indirect {
 				p = ptrToPtr(p)
 			}
-			bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+			bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 			if err != nil {
 				return nil, err
 			}
@@ -2718,7 +2718,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -2762,7 +2762,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -2807,7 +2807,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.EscapedKey...)
-				bb, err := appendMarshalJSON(code, b, iface, true)
+				bb, err := appendMarshalJSON(ctx, code, b, iface, true)
 				if err != nil {
 					return nil, err
 				}
@@ -2847,7 +2847,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -2887,7 +2887,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				code = code.NextField
 			} else {
 				b = append(b, code.EscapedKey...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -3628,7 +3628,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 && code.Nilcheck {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -3652,7 +3652,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 				break
 			}
 			b = append(b, code.EscapedKey...)
-			bb, err := appendMarshalJSON(code, b, iface, true)
+			bb, err := appendMarshalJSON(ctx, code, b, iface, true)
 			if err != nil {
 				return nil, err
 			}
@@ -3665,7 +3665,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			if p == 0 {
 				b = appendNull(b)
 			} else {
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}
@@ -3678,7 +3678,7 @@ func Run(ctx *encoder.RuntimeContext, b []byte, codeSet *encoder.OpcodeSet, opt 
 			p = ptrToNPtr(p+code.Offset, code.PtrNum)
 			if p != 0 {
 				b = append(b, code.EscapedKey...)
-				bb, err := appendMarshalJSON(code, b, ptrToInterface(code, p), true)
+				bb, err := appendMarshalJSON(ctx, code, b, ptrToInterface(code, p), true)
 				if err != nil {
 					return nil, err
 				}

--- a/json_test.go
+++ b/json_test.go
@@ -80,6 +80,12 @@ func TestCompact(t *testing.T) {
 			t.Errorf("Compact(%#q) = %#q, want %#q", tt.indent, s, tt.compact)
 		}
 	}
+	t.Run("invalid", func(t *testing.T) {
+		buf.Reset()
+		if err := json.Compact(&buf, []byte(`invalid`)); err == nil {
+			t.Fatal(err)
+		}
+	})
 }
 
 func TestCompactSeparators(t *testing.T) {

--- a/json_test.go
+++ b/json_test.go
@@ -197,7 +197,7 @@ func TestIndentErrors(t *testing.T) {
 		buf := bytes.NewBuffer(slice)
 		if err := json.Indent(buf, []uint8(tt.in), "", ""); err != nil {
 			if !reflect.DeepEqual(err, tt.err) {
-				t.Errorf("#%d: Indent: %#v", i, err)
+				t.Errorf("#%d: Indent: expected %#v but got %#v", i, tt.err, err)
 				continue
 			}
 		}

--- a/json_test.go
+++ b/json_test.go
@@ -2,6 +2,7 @@ package json_test
 
 import (
 	"bytes"
+	stdjson "encoding/json"
 	"math"
 	"math/rand"
 	"reflect"
@@ -81,9 +82,34 @@ func TestCompact(t *testing.T) {
 		}
 	}
 	t.Run("invalid", func(t *testing.T) {
-		buf.Reset()
-		if err := json.Compact(&buf, []byte(`invalid`)); err == nil {
-			t.Fatal(err)
+		for _, src := range []string{
+			`invalid`,
+			`}`,
+			`]`,
+			`{"a":1}}`,
+			`{"a" 1}`,
+			`{"a": 1 "b": 2}`,
+			`["a" "b"]`,
+			`"\`,
+			`{"a":"\\""}`,
+			`tr`,
+			`{"a": tru, "b": 1}`,
+			`fal`,
+			`{"a": fals, "b": 1}`,
+			`nu`,
+			`{"a": nul, "b": 1}`,
+			`1.234.567`,
+			`[nul]`,
+			`{}   1`,
+		} {
+			buf.Reset()
+			if err := stdjson.Compact(&buf, []byte(src)); err == nil {
+				t.Fatal("invalid test case")
+			}
+			buf.Reset()
+			if err := json.Compact(&buf, []byte(src)); err == nil {
+				t.Fatalf("%q: expected error", src)
+			}
 		}
 	})
 }
@@ -125,6 +151,37 @@ func TestIndent(t *testing.T) {
 			t.Errorf("Indent(%#q) = %#q, want %#q", tt.compact, s, tt.indent)
 		}
 	}
+	t.Run("invalid", func(t *testing.T) {
+		for _, src := range []string{
+			`invalid`,
+			`}`,
+			`]`,
+			`{"a":1}}`,
+			`{"a" 1}`,
+			`{"a": 1 "b": 2}`,
+			`["a" "b"]`,
+			`"\`,
+			`{"a":"\\""}`,
+			`tr`,
+			`{"a": tru, "b": 1}`,
+			`fal`,
+			`{"a": fals, "b": 1}`,
+			`nu`,
+			`{"a": nul, "b": 1}`,
+			`1.234.567`,
+			`[nul]`,
+			`{}   1`,
+		} {
+			buf.Reset()
+			if err := stdjson.Indent(&buf, []byte(src), "", " "); err == nil {
+				t.Fatal("invalid test case")
+			}
+			buf.Reset()
+			if err := json.Indent(&buf, []byte(src), "", " "); err == nil {
+				t.Fatalf("%q: expected error", src)
+			}
+		}
+	})
 }
 
 // Tests of a large random structure.

